### PR TITLE
Add a documentation page for `patronictl`

### PR DESCRIPTION
**Pull request by barthisrael**
This PR introduces a documentation page for `patronictl` application.

We adopted a top-down approach when writing this document. We start by describing the outer most parts, and then keep writing new sections that specialize the knowledge.

We basically added a section called `patronictl` to the left menu. Inside that section we created a page with this structure:

- `patronictl`: describes what it is
    - `Configuraiton`: how to configure `patronictl`
    - `Usage`: how to use the CLI. Inside this section, there are subsections for each of the subcommands exposed by `patronictl`, and each of them are described using the following subsubsections:
        - `Synopsis`: syntax of the command and its positional and optional arguments
        - `Description`: a description of what the command does
        - `Parameters`: a detailed description of the arguments and how to use them
        - `Examples`: one or more examples of execution of the command

References: PAT-200.

### CODE REVIEW

**Comment by hughcapet** [comment]
one thought straight away - should we add references to the new page for the existing occurrences of patronictl in the other parts of the docs?

**Comment by barthisrael** [comment]
> one thought straight away - should we add references to the new page for the existing occurrences of patronictl in the other parts of the docs?

Hmm, good point, I guess so.

**Comment by barthisrael** [comment]
I just submitted a commit that will allow any cross-reference to sections/subsections/subsubsections under `patronictl` page.

However, there is a behavior that I was not able to implement properly regarding the `patronictl subcommands` subsubsections: the first command, `patronictl dsn`, doesn't use the same anchor name format than others. I mean:

* I have these labes for `patronictl dsn` subsubsections:
    * `.. _patronictl_dsn_synopsis:`
    * `.. _patronictl_dsn_description:`
    * `.. _patronictl_dsn_parameters:`
    * `.. _patronictl_dsn_examples:`

But the HTML anchor IDs are `#synopsis`, `#description`, `#parameters`, and `#examples`, respectively.

That seems to occur only for the first occurrences of the duplicated subsubections names.

From the second subcommand onward (`patronictl edit-config` onward), I get anchor IDs named as I would expect: `#patronictl-edit-config-synopsis`, `#patronictl-edit-config-description`, `#patronictl-edit-config-parameters`, `#patronictl-edit-config-examples`, respectively.

Any idea on how to fix that on Sphinx, and make it use the same anchor ID naming convention for the first occurrences of duplicated subsubsections names too?

**Comment by barthisrael** [comment]
_I just noticed that each commit is triggering a new workflow run. I'll submit the next ones in a single batch_

**Comment by barthisrael** [comment]
I'm done with the commits. My only pending issues is the one described here (if that's an issue): https://github.com/zalando/patroni/pull/2874#issuecomment-1727941755.

I was not able to find a built-in way or resolving that. Do you happen to know any?

File: docs/patronictl.rst [hunk 1]
```diff
@@ -0,0 +1,1963 @@
...
+Parameters
+""""""""""
+
+``CLUSTER_NAME``
+    Name of the Patroni cluster.
+
+    If not given, ``patronictl`` will attempt to fetch that from the ``scope`` configuration, if it exists.
+
+``--group``
+    Change dynamic configuration of the given Citus group.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:205]:
```suggestion
    Change dynamic configuration of the given Citus group.
    
    If not given, ``patronictl`` will attempt to fetch that from the ``citus.group`` configuration, if it exists.
```

File: docs/patronictl.rst [hunk 2]
```diff
@@ -0,0 +1,1963 @@
...
+
+``CLUSTER_NAME``
+    Name of the Patroni cluster.
+
+    If not given, ``patronictl`` will attempt to fetch that from the ``scope`` configuration, if it exists.
+
+``--group``
+    Pause the given Citus group.
+
+    ``CITUS_GROUP`` is the ID of the Citus group.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:912]:
```suggestion
    ``CITUS_GROUP`` is the ID of the Citus group.
    
    If not given, ``patronictl`` will attempt to fetch that from the ``citus.group`` configuration, if it exists.
```

File: docs/patronictl.rst [hunk 3]
```diff
@@ -0,0 +1,1963 @@
...
+
+``CLUSTER_NAME``
+    Name of the Patroni cluster.
+
+    If not given, ``patronictl`` will attempt to fetch that from the ``scope`` configuration, if it exists.
+
+``--group``
+    Resume the given Citus group.
+
+    ``CITUS_GROUP`` is the ID of the Citus group.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:1528]:
```suggestion
    ``CITUS_GROUP`` is the ID of the Citus group.
    
    If not given, ``patronictl`` will attempt to fetch that from the ``citus.group`` configuration, if it exists.
```

File: docs/patronictl.rst [hunk 4]
```diff
@@ -0,0 +1,1963 @@
...
+
+``CLUSTER_NAME``
+    Name of the Patroni cluster.
+
+    If not given, ``patronictl`` will attempt to fetch that from the ``scope`` configuration, if it exists.
+
+``--group``
+    Show history of events from the given Citus group.
+
+    ``CITUS_GROUP`` is the ID of the Citus group.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:574]:
```suggestion
    ``CITUS_GROUP`` is the ID of the Citus group.
    
    If not given, ``patronictl`` will attempt to fetch that from the ``citus.group`` configuration, if it exists.
```

File: docs/patronictl.rst [hunk 5]
```diff
@@ -0,0 +1,1963 @@
...
+
+``CLUSTER_NAME``
+    Name of the Patroni cluster.
+
+    If not given, ``patronictl`` will attempt to fetch that from the ``scope`` configuration, if it exists.
+
+``--group``
+    Show dynamic configuration of the given Citus group.
+
+    ``CITUS_GROUP`` is the ID of the Citus group.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:1582]:
```suggestion
    ``CITUS_GROUP`` is the ID of the Citus group.
    
    If not given, ``patronictl`` will attempt to fetch that from the ``citus.group`` configuration, if it exists.
```

File: docs/patronictl.rst [hunk 6]
```diff
@@ -0,0 +1,1963 @@
...
+    Perform a failover in the given Citus group.
+
+    ``CITUS_GROUP`` is the ID of the Citus group.
+
+``--leader`` / ``--primary`` / ``--master``
+    Indicate who is the expected leader at failover time.
+
+    If given, a switchover is performed instead of a failover.
+
+    ``LEADER_NAME`` should match the name of the current leader in the cluster.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:364]: It is getting deprecated, should we still document it?

File: docs/patronictl.rst [hunk 7]
```diff
@@ -0,0 +1,1963 @@
...
+Description
+"""""""""""
+
+``patronictl failover`` performs a manual failover in the cluster.
+
+It is designed to be used when the cluster is not healthy, e.g.:
+
+- There is no leader; or
+- There is no synchronous standby available in a synchronous cluster.
+
```
Comment (CyberDem0n) [inline docs/patronictl.rst:337]:
```suggestion

It also allows to fail over to asynchronous node is synchronous mode is enabled.
```

File: docs/patronictl.rst [hunk 8]
```diff
@@ -0,0 +1,1963 @@
...
+      [ --group CITUS_GROUP ]
+      [ { -r | --role } { leader | primary | standby-leader | replica | standby | any | master } ]
+      [ --force ]
+
+.. _patronictl_reload_description:
+
+Description
+"""""""""""
+
+``patronictl reload`` requests a reload of local configuration for one or more Patroni members.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:1221]: It also triggers `pg_ctl reload` even if nothing have been changed.

File: docs/patronictl.rst [hunk 9]
```diff
@@ -0,0 +1,1963 @@
...
+
+Description
+"""""""""""
+
+``patronictl switchover`` performs a switchover in the cluster.
+
+It is designed to be used when the cluster is healthy, e.g.:
+
+- There is a leader; or
+- There are synchronous standbys available in a synchronous cluster.
```
Comment (CyberDem0n) [inline docs/patronictl.rst:1634]: I read it as `we could have synchronous standbys but no leader`, which is wrong.

**Review by CyberDem0n — COMMENTED** [review]
Looks super nice!
I left a few comments:
- maybe we can sometimes use different style for code blocks (mainly for patronictl edit-config)
- maybe we should prepend all commands with `$` character to emphasize that this is is supposed to be executed from shell
- for come commands Citus group could be taken from the configuration file

File: docs/patronictl.rst (hunk 6 above)
Comment (barthisrael) [reply docs/patronictl.rst:366]: I guess so. I thought of adding a deprecation notice. We can remove it later, once we remove this altogether.

File: docs/patronictl.rst (hunk 9 above)
Comment (barthisrael) [reply docs/patronictl.rst:1644]: Removed the "or". I guess that's enough, right?

**Comment by barthisrael** [comment]
I guess I covered all points.
Thanks for your thoughts!

**Comment by CyberDem0n** [comment]
> But the HTML anchor IDs are #synopsis, #description, #parameters, and #examples, respectively.

@barthisrael https://github.com/GeeTransit/sphinx-better-subsection seems to do what you want.
I wonder though if it will introduce some side effects on other parts of docs and how critical they would be.

**Comment by barthisrael** [comment]
> @barthisrael https://github.com/GeeTransit/sphinx-better-subsection seems to do what you want.
> I wonder though if it will introduce some side effects on other parts of docs and how critical they would be.

Ah, you found something :)

Right! And I wonder how much of a problem the described issue would be:

* If someone tries to infere the links for `patronictl dsn` subsections based on the pattern of the other commands, they will reach the head of the document;
* If someone saved the `#synopsis` link for `patronictl dsn` and we introduce a newer command in the future and put it before `patronictl dsn`, the link will go to the new command.

IAC I think the former is not much of an issue, and the second is very unlikely to occur.

Maybe we should just leave it as it is?

**Comment by CyberDem0n** [comment]
The problem is not with `patronictl.rst`, but with the rest of documentation.
If someone was referring to anchors they might stop working. At the moment there are ~80 such places.
From another side if we change titles it also changes anchors and links also break.

**Comment by barthisrael** [comment]
> The problem is not with `patronictl.rst`, but with the rest of documentation. If someone was referring to anchors they might stop working. At the moment there are ~80 such places. From another side if we change titles it also changes anchors and links also break.

Ah, yes, I got it when you mentioned about the extension.

I was just wondering that by keeping the current behavior we might not cause much problems to people consulting patronictl doc page other than what I described above.

File: docs/patronictl.rst [hunk 10]
```diff
@@ -0,0 +1,1975 @@
...
+Patroni has a command-line interface named ``patronictl``, which is used basically to interact with Patroni's REST API and with the DCS. It is intended to make it easier to perform operations in the cluster, and can easily be used by humans or scripts.
+
+.. _patronictl_configuration:
+
+Configuration
+-------------
+
+``patronictl`` uses 3 sections of the configuration:
+
+- **ctl**: how to authenticate against the REST API server, and how to validate the server identity;
```
[the remark this answers is not in the document]
Comment (barthisrael) [reply docs/patronictl.rst:15]: Good point!

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**